### PR TITLE
wr-66: Add password comparison upon Sign In

### PR DESCRIPTION
**Pull request by kovAnya**
Do I need to move InvalidCredentialsError to shared\src\libs\exceptions\invalid-credentials-error\invalid-credentials-error.exception.ts?

### CODE REVIEW

File: backend/src/packages/users/user.service.ts [hunk 1]
```diff
@@ -51,6 +52,21 @@ class UserService implements IService {
...
+
+    const user = await this.userRepository.findByEmail(email);
+
+    return user
+      ? await this.encrypt.checkPassword(
+          password,
+          user.toNewObject().passwordHash,
+        )
+      : false;
+  }
```
Comment (andanf-e) [inline backend/src/packages/users/user.service.ts]: And create new static getter `privateData` upon user entity that returns `passwordHash` and `passwordSalt`
```suggestion
  public async findPrivateData(id: number): Promise<UserPrivateData | null> {
    const user = await this.userRepository.find(id);

    if (!user) {
      return null;
    }

    return user.privateData;
  }
```

File: backend/src/packages/auth/auth.service.ts [hunk 2]
```diff
@@ -22,9 +22,14 @@ class AuthService {
 
     const user = await this.userService.findByEmail(email);
 
-    if (!user) {
-      throw new UserNotFoundError();
```
Comment (andanf-e) [inline backend/src/packages/auth/auth.service.ts:26]: Turn back user exists assertion

File: backend/src/packages/auth/auth.service.ts [hunk 3]
```diff
@@ -22,9 +22,14 @@ class AuthService {
...
+    const checkUserPassword = user
+      ? await this.userService.checkUserPassword(userSignInDto)
+      : null;
+
+    if (!user || !checkUserPassword) {
+      throw new InvalidCredentialsError();
```
Comment (andanf-e) [inline backend/src/packages/auth/auth.service.ts]:
```suggestion
     const userPrivateData = (await this.userService.findPrivateData(
      user.id,
    )) as UserPrivateData;

    const hasSamePassword = await this.encrypt.compare({
      data: password,
      salt: userPrivateData.passwordSalt,
      passwordHash: userPrivateData.passwordHash,
    });

    if (!hasSamePassword) {
        throw new InvalidCredentialsError();
    }
```

File: backend/src/libs/packages/encrypt/encrypt.package.ts [hunk 4]
```diff
@@ -10,6 +10,13 @@ class Encrypt implements IEncrypt {
   public async encrypt(password: string, salt: string): Promise<string> {
     return await hash(password, salt);
   }
+
+  public async checkPassword(
+    password: string,
+    hashPassword: string,
+  ): Promise<boolean> {
+    return await compare(password, hashPassword);
+  }
```
Comment (andanf-e) [inline backend/src/libs/packages/encrypt/encrypt.package.ts]:
```suggestion
  public async compare({
    passwordToCompare,
    salt,
    passwordHash,
  }: {
    passwordToCompare: string;
    salt: string;
    passwordHash: string;
  }): Promise<boolean> {
    const hashToCompare = await this.encrypt(data, salt);

    return hashToCompare === passwordHash;
  }
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: backend/src/packages/auth/auth.service.ts (hunk 2 above)
Comment (kovAnya) [reply backend/src/packages/auth/auth.service.ts:26]: @andanf-e So the user will know if the email is wrong before comparing the password? I assumed that the user just needs to know that the email and password match, not that the email doesn't exist. Or should I change the UserNotFoundError's message?

Comment (mhrechkin) [reply backend/src/packages/auth/auth.service.ts:26]: If user not found, the error should be shown before password comparison.

**Review by mhrechkin — CHANGES_REQUESTED** [review]

File: backend/src/packages/users/user.service.ts [hunk 5]
```diff
@@ -51,6 +52,14 @@ class UserService implements IService {
     };
   }
 
+  public async findPrivateData(id: number): Promise<UserPrivateData | null> {
+    const user = await this.userRepository.find(id);
+    if (!user) {
+      return null;
+    }
```
Comment (andanf-e) [inline backend/src/packages/users/user.service.ts:60]:
```suggestion

        if (!user) {
      return null;
    }
    
```

File: backend/src/packages/auth/auth.service.ts [hunk 6]
```diff
@@ -8,23 +12,40 @@ import {
...
+      user.id,
+    )) as UserPrivateData;
+    const hasSamePassword = await this.encrypt.compare({
+      passwordToCompare: password,
+      salt: userPrivateData.passwordSalt,
+      passwordHash: userPrivateData.passwordHash,
+    });
+    if (!hasSamePassword) {
+      throw new InvalidCredentialsError();
+    }
```
Comment (andanf-e) [inline backend/src/packages/auth/auth.service.ts]:
```suggestion
    const userPrivateData = (await this.userService.findPrivateData(
      user.id,
    )) as UserPrivateData;
    
    const hasSamePassword = await this.encrypt.compare({
      passwordToCompare: password,
      salt: userPrivateData.passwordSalt,
      passwordHash: userPrivateData.passwordHash,
    });
    
    if (!hasSamePassword) {
      throw new InvalidCredentialsError();
    }
```

File: backend/src/libs/packages/encrypt/encrypt.package.ts [hunk 7]
```diff
@@ -10,6 +10,19 @@ class Encrypt implements IEncrypt {
...
+    passwordToCompare,
+    salt,
+    passwordHash,
+  }: {
+    passwordToCompare: string;
+    salt: string;
+    passwordHash: string;
+  }): Promise<boolean> {
+    const hashToCompare = await this.encrypt(passwordToCompare, salt);
+    return hashToCompare === passwordHash;
```
Comment (andanf-e) [inline backend/src/libs/packages/encrypt/encrypt.package.ts:25]: Without any spacing dividers it's not the easiest thing to read
```suggestion

    return hashToCompare === passwordHash;
```

Comment (kovAnya) [reply backend/src/libs/packages/encrypt/encrypt.package.ts:25]: ok, got it

**Review by andanf-e — DISMISSED** [review]

File: backend/src/packages/auth/auth.service.ts [hunk 8]
```diff
@@ -8,23 +12,42 @@ import {
...
-    const { email } = userSignInDto;
+    const { email, password } = userSignInDto;
 
     const user = await this.userService.findByEmail(email);
-
     if (!user) {
       throw new UserNotFoundError();
     }
+
+    const userPrivateData = (await this.userService.findPrivateData(
```
Comment (PablloZz) [inline backend/src/packages/auth/auth.service.ts:37]: Would it better to use destructuring here?

File: backend/src/packages/auth/auth.service.ts [hunk 9]
```diff
@@ -8,23 +12,42 @@ import {
...
+      user.id,
+    )) as UserPrivateData;
```
Comment (PablloZz) [inline backend/src/packages/auth/auth.service.ts:39]: Would it be better to use destructuring here?

Comment (kovAnya) [reply backend/src/packages/auth/auth.service.ts:39]: I believe that it won't be useful a lot. This statement isn't too complex, is it?

Comment (andanf-e) [reply backend/src/packages/auth/auth.service.ts:39]: Let's keep it as it is - like this we're grouping what values is internal, not from request payload

**Review by andanf-e — DISMISSED** [review]

**Review by andanf-e — APPROVED** [review]

**Review by NazarKovtun1 — APPROVED** [review]

**Review by mhrechkin — APPROVED** [review]

**Pull request merged**